### PR TITLE
Variety of changes: null literals, throttle writer, GC, etc.

### PR DESCRIPTION
**Pull request by kosak**
1. Allow the user to configure multiple "null literals" rather than one,
   and do so globally or on a per-column basis.
2. In the concurrent case, throttle the writer so it can't get too far
   ahead of the reader.
3. Be much more careful about holding local variables to the
   DenseStorageReader. This fixes a problem where locals were holding
   on to a prefix of the linked list, preventing it from being GC'ed
   even when it otherwise would have an opportunity to do so.
4. Change the DenseStorageReader interface so you can copy() it.
   Previously I had this odd programming interface where you had to
   make all your DenseStorageReaders up front, from the Writer, before
   any other changes were made. Now make a DenseStorageReader and Writer
   pair up front, but you can clone (via a method called copy()) the
   DenseStorageWriter any time you want.

### CODE REVIEW

File: src/main/java/io/deephaven/csv/util/Util.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+public class Util {
+    /** Call Object.wait() but suppress the need to deal with checked InterruptedExceptions. */
+    public static void catchyWait(Object o) {
+        try {
+            o.wait();
+        } catch (InterruptedException ie) {
+            throw new RuntimeException(
+                    "Thread interrupted: probably cancelled in the CsvReader class due to some other exception.");
+        }
```
Comment (devinrsmith) [inline src/main/java/io/deephaven/csv/util/Util.java:11]: IMO, we should be trying to push InterruptedException's as far up the stack as possible. It lets callers know "this call may block", and gives higher level logic (which may have more context) the ability to more gracefully deal with it.

I'm not sure it's worth piping through today, but I think we may need to revisit this in the future.

**Review by devinrsmith — COMMENTED** [review]
This is dense. I'm not sure I understand all that's going on. I made some high level comments. I'm going to spend a little more time looking at this and check out the code.

File: src/main/java/io/deephaven/csv/reading/ParseInputToDenseStorage.java [hunk 2]
```diff
@@ -102,12 +104,20 @@ public RowAppender(final byte[][] optionalFirstDataRow, final CellGrabber grabbe
...
-                final byte[] nullValueBytes = nvl.getBytes(StandardCharsets.UTF_8);
-                nullSlice = new ByteSlice(nullValueBytes, 0, nullValueBytes.length);
-            } else {
-                nullSlice = null;
+            // In case we need to null-fill short columns, we need a suitable null value literal for each column.
+            // Note that the caller can specify zero, one, or more null value literals, and they can do so on a
+            // per-column basis. If there is no suitable null value literal for a column, we calculate a null here.
+            // Otherwise, if there are one or more suitable null value literals for a column, we arbitrarily take the
+            // first one. In this latter case it doesn't matter to us which configured null value literal we use
+            // for a column, because they should all be interchangeable.
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/reading/ParseInputToDenseStorage.java]: I added a detailed comment (and changed the code slightly).

File: src/main/java/io/deephaven/csv/CsvSpecs.java [hunk 3]
```diff
@@ -273,23 +272,23 @@ public List<Parser<?>> parsers() {
...
 
     /**
-     * See {@link Builder#nullValueLiteral}.
+     * See {@link Builder#nullValueLiterals}.
      */
     @Default
     @Nullable
-    public String nullValueLiteral() {
-        return "";
+    public String[] nullValueLiterals() {
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/CsvSpecs.java:279]: Updated to List&lt;String&gt;

File: src/main/java/io/deephaven/csv/densestorage/DenseStorageConstants.java [hunk 4]
```diff
@@ -31,4 +33,9 @@ public class DenseStorageConstants {
      * rather its own byte array.
      */
     public static final int LARGE_BYTE_ARRAY_SENTINEL = -1;
+    /**
+     * The maximum number of data blocks that we allow to go unobserved before the blocking the QueueWriter. This is
+     * only used when {@link CsvSpecs#concurrent()} is true.
+     */
+    public static final int MAX_UNOBSERVED_BLOCKS = 4;
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/densestorage/DenseStorageConstants.java:40]: I don't really think so. I just picked a small number seemed to leave a little elbow room. I don't think this is a parameter that will make any performance difference if made larger. Put another way, I don't think there's any benefit to the writer getting way ahead of the readers.

File: src/main/java/io/deephaven/csv/densestorage/QueueNode.java [hunk 5]
```diff
@@ -23,4 +29,22 @@ public QueueNode(TARRAY data, int begin, int end, boolean isLast) {
         this.isLast = isLast;
         this.next = null;
     }
+
+    public synchronized void setNext(QueueNode<TARRAY> next) {
+        this.next = next;
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/densestorage/QueueNode.java:34]: I have combined the "QueueState" functionality into QueueNode, having it access the semaphore directly. Also it does the check you suggest in side there.

File: src/main/java/io/deephaven/csv/densestorage/QueueState.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package io.deephaven.csv.densestorage;
+
+import io.deephaven.csv.util.Util;
+
+/**
+ * This class is used to maintain a statistic of the number of unread blocks (blocks not observed by any reader). We use
+ * this so as to prevent the writer from getting too far ahead of the slowest reader.
+ */
+public class QueueState {
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/densestorage/QueueState.java:9]: done, thanks

File: src/main/java/io/deephaven/csv/densestorage/QueueState.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+     * The number of blocks that have been appended to the queue but never read by any QueueReader.
+     */
+    private int numUnobservedBlocks;
+
+    public QueueState(int maxUnobservedBlocks) {
+        this.maxUnobservedBlocks = maxUnobservedBlocks;
+        this.numUnobservedBlocks = 0;
+    }
+
+    public synchronized void noteBlockAdded() {
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/densestorage/QueueState.java:27]: done, thanks

**Review by devinrsmith — APPROVED** [review]

**Pull request merged**